README.md

```diff
@@ -2,31 +2,359 @@
+constructor methods.
+
+## Implicit Parameter
+
+The variable `this` is referred to as an *implicit parameter*
+because the instance method does not *explicitly* declare
+it as a parameter.   An implicit parameter is not defined
+as part of the method signature, but it is implied by the class
+declaring an instance method (i.e. a non-static method).
+
+Since Java uses `this` as a keyword for the implicit parameter's name,
+we can't define a variable named `this`.  We also can't assign
+`this` a value since that is done automatically when an instance
+method is called.
+
+## Debugging Instance Methods  (Code Along)
+
+Let's use the debugger to step into an instance method and see how it
+changes the state of the object referenced by `this`. 
+
+So far we've mostly used IntelliJ's step-over button while debugging.
+If the current line of code contains a method call, the step-over button
+executes the method call and then proceeds to the next line of code.
+
+![step over](https://curriculum-content.s3.amazonaws.com/6676/java-mod2-strings/step_over.png)
+
+However, we would like to "step-into" the instance methods to
+see how `this` is used to access the instance variables. 
+
+![step into](https://curriculum-content.s3.amazonaws.com/6676/java-mod2-strings/step_into.png)
+
+
+1. Set a breakpoint at the line of code `fido.giveTreat();` and launch the debugger.    
+   ![breakpoint](https://curriculum-content.s3.amazonaws.com/6676/java-mod2-strings/fido_breakpoint.png)
+2. Switch to the Java Visualizer view. The visualizer shows the current object state *prior* to
+   executing `fido.giveTreat()`.    
+   ![dog state](https://curriculum-content.s3.amazonaws.com/6676/java-mod2-strings/dog_this_0.png)
+3. Press step-into to execute the method call `fido.giveTreat();`.
+   Notice the method `getTreat()` is added to the call stack and  `this`
+   references the same object as `fido`. Since step-into was pressed,
+   the debugger stops at the first line of code in the `getTreat()` method.    
+   ![fido debug](https://curriculum-content.s3.amazonaws.com/6676/java-mod2-strings/debugger_this_fido.png)
+4. You can press either step-over  or step-into to execute the two assignment
+   statements in the `giveTreat()` method since neither makes a method call.
+   Keep in mind that the compiler evolves the code to use `this` to access the instance variables:    
+    
+   | Programmer writes:      | Compiler generates:          |
+   |-------------------------|------------------------------|
+   | `isWaggingTail = true;` | `this.isWaggingTail = true;` |
+   | `weight++;`             | `this.weight++;`             |
+   
+   Step twice to execute the two assignment statements. The object referenced by `this` is updated:     
+   ![fido debug update state](https://curriculum-content.s3.amazonaws.com/6676/java-mod2-strings/debugger_this_fido2.png)
+   
+5. Step again to complete the `getTreat()` method and return to the `main` method.
```

Comment:
Should "Step again" be "Step-over again"?

Reply:
yes, updated